### PR TITLE
sys/auto_init: added mechanism for initialization of network devices

### PR DESCRIPTION
**Pull request by haukepetersen**
This PR introduces a mechanism, that allows for auto-configuration of network interfaces. The configuration is normally done by each board (of course it is interesting only for those that have network interfaces), but can be overwritten by the application (may you want a different MAC layer...).

### CODE REVIEW

File: boards/iot-lab_M3/auto_init_ng_netif/netif_board.c [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+ */
+static ng_at86rf2xx_t radio;
+
+/**
+ * @brief   Stack for the MAC layer thread
+ */
+static char nomac_stack[MAC_STACKSIZE];
+
+
+void auto_init_ng_netif(void)
```
Comment (authmillenon) [inline boards/iot-lab_M3/auto_init_ng_netif/netif_board.c]: Either name this function or the module `ng_auto_init_netif` or just `auto_init_netif`. It's worth enough that `ng_` has become an infix for `MODULE_` macros and `net_` doxygen groups. For the future switch-over I'd like to have as few corner-cases as possible. Currently there is (in sed syntax) `\<ng_`, `\<NG_`, `\<MODULE_NG` and `\<net_ng`.

File: boards/iot-lab_M3/auto_init_ng_netif/netif_board.c [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+/**
+ * @brief   Define stack parameters for the MAC layer thread
+ * @{
+ */
+#define MAC_STACKSIZE           (KERNEL_CONF_STACKSIZE_DEFAULT)
+#define MAC_PRIO                (0U)
```
Comment (OlegHahm) [inline boards/iot-lab_M3/auto_init_ng_netif/netif_board.c]: Is really required (and wise) to give the MAC protocol the highest possible priority?

File: boards/iot-lab_M3/auto_init_ng_netif/netif_board.c (hunk 1 above)
Comment (haukepetersen) [inline boards/iot-lab_M3/auto_init_ng_netif/netif_board.c]: but this does syntactical not make sense: the notion is: `auto_init_MODULE_NAME`, and the module is called `ng_netif`...

File: boards/iot-lab_M3/auto_init_ng_netif/netif_board.c (hunk 2 above)
Comment (haukepetersen) [inline boards/iot-lab_M3/auto_init_ng_netif/netif_board.c]: probably not. I will fix this to a less extreme value

File: Makefile.include [hunk 3]
```diff
@@ -16,6 +16,9 @@ RIOTPROJECT := $(abspath $(RIOTPROJECT))
 # Path to the current directory relative to the git root
 BUILDRELPATH ?= $(shell git rev-parse --show-prefix)
 
+APPDIR ?= $(CURDIR)
+APPDIR := $(abspath $(APPDIR))/
+
 BINDIRBASE ?= $(CURDIR)/bin
```
Comment (authmillenon) [inline Makefile.include]: optional: Maybe set this to APPDIR too then

File: sys/auto_init/auto_init.c [hunk 4]
```diff
@@ -304,4 +309,12 @@ void auto_init(void)
     DEBUG("Auto init UDP module.\n");
     ng_udp_init();
 #endif
+#ifdef MODULE_NG_NETIF
+    DEBUG("Auto init netif module.\n");
+    ng_netif_init();
```
Comment (authmillenon) [inline sys/auto_init/auto_init.c]: This function does not need to be called. I think I have to put this into `TEST_SUITES` context :D

Comment (authmillenon) [inline sys/auto_init/auto_init.c]: (`ng_netif_init()` just zeros the `netif` array, which is zero anyway on startup)

**Comment by authmillenon** [comment]
αpart from the comments above and the fact, that there is still room for optimization IMHO (as always) I'm fine with this PR.

**Comment by haukepetersen** [comment]
addressed comments. Should I squash?

**Comment by authmillenon** [comment]
Yes, please (This is an ACK, if travis is happy)

File: tests/driver_at86rf2xx/auto_init_ng_netif/netif_app.c [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+#ifndef ATRF_SPI_SPEED
+#define ATRF_SPI_SPEED          (SPI_SPEED_5MHZ)
+#endif
+
+/**
+ * @brief   MAC layer stack configuration
+ * @{
+ */
+#define STACKSIZE               (KERNEL_CONF_STACKSIZE_MAIN)
+#define PRIO                    (0)
```
Comment (authmillenon) [inline tests/driver_at86rf2xx/auto_init_ng_netif/netif_app.c]: Priority needs to be adjusted

File: tests/driver_xbee/auto_init_ng_netif/netif_app.c [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+ *          when you buy them
+ */
+#define XBEE_BAUDRATE           (9600U)
+
+/**
+ * @brief   MAC layer stack configuration
+ * @{
+ */
+#define STACKSIZE               (KERNEL_CONF_STACKSIZE_DEFAULT)
+#define PRIO                    (0)
```
Comment (authmillenon) [inline tests/driver_xbee/auto_init_ng_netif/netif_app.c]: here too

**Comment by authmillenon** [comment]
Please address my last to comments also

File: tests/driver_xbee/auto_init_ng_netif/netif_app.c (hunk 6 above)
Comment (haukepetersen) [inline tests/driver_xbee/auto_init_ng_netif/netif_app.c]: in the test cases the prio does not hurt - as here will never be 'unexpected' other threads...

**Comment by haukepetersen** [comment]
squashed. Wontfix your last comments as in my opinion they do not hurt and changing the prio would not change anything...

**Comment by authmillenon** [comment]
Alright then … re-ACK

**Pull request merged**